core/src/test/scala/kafka/server/ForwardingManagerMetricsTest.scala

```diff
@@ -0,0 +1,101 @@
+      new MetricName("QueueTimeMs.p99", expectedGroup, "", Collections.emptyMap()),
+      new MetricName("QueueTimeMs.p999", expectedGroup, "", Collections.emptyMap()),
+      new MetricName("QueueLength", expectedGroup, "", Collections.emptyMap()),
+      new MetricName("RemoteTimeMs.p99", expectedGroup, "", Collections.emptyMap()),
+      new MetricName("RemoteTimeMs.p999", expectedGroup, "", Collections.emptyMap())
+    )
+
+    Using(ForwardingManagerMetrics(metrics)) { _ =>
+      val metricsMap = metrics.metrics().asScala.filter { case (name, _) => name.group == expectedGroup }
+      assertEquals(metricsMap.size, expectedMetrics.size)
+      metricsMap.foreach { case (name, _) =>
+        assertTrue(expectedMetrics.contains(name))
+      }
+    }
+
+    val metricsMap = metrics.metrics().asScala.filter { case (name, _) => name.group == expectedGroup }
+    assertEquals(0, metricsMap.size)
+  }
+
+  @Test
+  def testQueueTimeMs(): Unit = {
+    val metrics = new Metrics()
+
+    Using(ForwardingManagerMetrics(metrics)) { forwardingManagerMetrics =>
+      val queueTimeMsP99 = metrics.metrics().get(forwardingManagerMetrics.queueTimeMsHist.latencyP99Name)
+      val queueTimeMsP999 = metrics.metrics().get(forwardingManagerMetrics.queueTimeMsHist.latencyP999Name)
+      assertEquals(Double.NaN, queueTimeMsP99.metricValue.asInstanceOf[Double])
+      assertEquals(Double.NaN, queueTimeMsP999.metricValue.asInstanceOf[Double])
+      for(i <- 0 to 999) {
+        forwardingManagerMetrics.queueTimeMsHist.record(i)
+      }
+      assertEquals(990.0, queueTimeMsP99.metricValue.asInstanceOf[Double])
+      assertEquals(999.0, queueTimeMsP999.metricValue.asInstanceOf[Double])
+    }
+  }
+
+  @Test
+  def testQueueLength(): Unit = {
+    val metrics = new Metrics()
+
+    Using(ForwardingManagerMetrics(metrics)) { forwardingManagerMetrics =>
+      val queueLength = metrics.metrics().get(forwardingManagerMetrics.queueLengthName)
+      assertEquals(0, queueLength.metricValue.asInstanceOf[Int])
+      forwardingManagerMetrics.queueLength.getAndIncrement()
+      assertEquals(1, queueLength.metricValue.asInstanceOf[Int])
+    }
+  }
+
+  @Test
+  def testRemoteTimeMs(): Unit = {
+    val metrics = new Metrics()
+
+    Using(ForwardingManagerMetrics(metrics)) { forwardingManagerMetrics =>
+      val remoteTimeMsP99 = metrics.metrics().get(forwardingManagerMetrics.remoteTimeMsHist.latencyP99Name)
+      val remoteTimeMsP999 = metrics.metrics().get(forwardingManagerMetrics.remoteTimeMsHist.latencyP999Name)
+      assertEquals(Double.NaN, remoteTimeMsP99.metricValue.asInstanceOf[Double])
+      assertEquals(Double.NaN, remoteTimeMsP999.metricValue.asInstanceOf[Double])
+      for (i <- 0 to 999) {
+        forwardingManagerMetrics.remoteTimeMsHist.record(i)
+      }
+      assertEquals(990.0, remoteTimeMsP99.metricValue.asInstanceOf[Double])
+      assertEquals(999.0, remoteTimeMsP999.metricValue.asInstanceOf[Double])
+    }
+  }
+
```

Comment:
A timeout test would be good here. Perhaps we could set the configuration for that test so that timeout happened quickly (is that configurable?)

Reply:
I think it would be best to add checks to existing `ForwardingManagerTest` tests that metrics are being appropriately updated when timeouts occur (and when requests are handled). My `ForwardingManagerMetricsTest` file contains unit tests that don't mock the `NodeToControllerChannelManager`.

Reply:
I can add a unit test showing that any larger recorded latencies will just be recorded as the timeout. I feel like also adding a timeout test (and also a test for normal behavior when `onComplete` is called rather than `onTimeout`) to `ForwardingManagerTest` instead of here would be nice too, since these are unit tests that don't actually build a request with a `ForwardingResponseHandler` or mock the `NodeToControllerChannelManager`.